### ISSUE TITLE
Please provide a FromIterator impl

### ISSUE THREAD

**Issue by joshtriplett** [BurntSushi/fst#86]
fst has all the necessary infrastructure for `FromIterator` to work; please consider implementing it, so that `collect` works, and in particular to make it easy to collect into `Result<Fst, ...>` or `Result<Set, ...>`.

Thank you!

**Comment by BurntSushi** [comment]
It's not clear to me that this is a good idea. Is there any precedent for implementing `FromIterator` for _fallible_ operations? As an example, [`Set::from_iter`](https://docs.rs/fst/0.3.5/fst/struct.Set.html#method.from_iter) already exists, but note that it returns a `Result`. That is, you cannot construct a `Set` from _any_ iterator. The keys must be lexicographically sorted. So as far as I can see, there are only two possible implementations for `FromIterator`:

* Consume the iterator into memory, sort the keys and then build the set.
* Panic if the iterator yields an unsorted key.

Neither option seems particularly good to me.

**Comment by joshtriplett** [comment]
Yes, there's precedent for fallible `FromIterator`; that's what allows `.collect::<Result<Vec<_>, SomeError>()`.

**Comment by BurntSushi** [comment]
Sure, but that only works because of a blanket impl on `Result` itself. Which means you can't implement `FromIterator<Vec<u8>> for Result<Fst<Vec<u8>>, fst::Error>`. Unless I'm missing something. If so, can you show me how? I don't see it.

Playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e997d571f41f88f3eba30660ee7b4429

Do you know of any `FromIterator` impls that are themselves fallible? I can't think of any off the top of my head in std. And a quick glance of the impls of `FromIterator` in std doesn't seem to show any. For example, if fallible `FromIterator` impls were allowed, then I'd expect to see a `FromIterator<u8> for String`, but I don't think such a thing can exist (without panicking or doing lossy decoding).
